### PR TITLE
[PRM-1418] Created GetFinancialDetailsService

### PR DESCRIPTION
**Pull request by JBlacklaws**

### CODE REVIEW

File: app/connectors/v3/getFinancialDetails/GetFinancialDetailsConnector.scala [hunk 1]
```diff
@@ -23,9 +23,9 @@ import uk.gov.hmrc.http.{HeaderCarrier, HttpClient}
...
 
-class FinancialDetailsConnector @Inject()(httpClient: HttpClient,
-                                          appConfig: AppConfig)
-                                         (implicit ec: ExecutionContext) {
+class GetFinancialDetailsConnector @Inject()(httpClient: HttpClient,
+                                             appConfig: AppConfig)
+                                            (implicit ec: ExecutionContext) {
 
   def getFinancialDetails(vatcUrl: String)(implicit hc: HeaderCarrier): Future[GetFinancialDetailsResponse] = {
     httpClient.GET[GetFinancialDetailsResponse](url = appConfig.getFinancialDetailsUrl + vatcUrl)
```
Comment (LewisMitchell) [inline app/connectors/v3/getFinancialDetails/GetFinancialDetailsConnector.scala:31]: Probably worth renaming the parameter here to `vrn`. There is some missing items from the url, see API 1811 spec

File: it/connectors/v3/getFinancialDetails/GetGetFinancialDetailsConnectorISpec.scala [hunk 2]
```diff
@@ -23,10 +23,10 @@ import play.api.test.Helpers.{await, defaultAwaitTimeout}
 import uk.gov.hmrc.http.HeaderCarrier
 import utils.{ETMPWiremock, IntegrationSpecCommonBase}
 
-class GetFinancialDetailsConnectorISpec extends IntegrationSpecCommonBase with ETMPWiremock with FeatureSwitching {
+class GetGetFinancialDetailsConnectorISpec extends IntegrationSpecCommonBase with ETMPWiremock with FeatureSwitching {
```
Comment (LewisMitchell) [inline it/connectors/v3/getFinancialDetails/GetGetFinancialDetailsConnectorISpec.scala:26]: Could you change the class name and file name back to `GetFinancialDetailsConnectorISpec`

File: it/services/GetFinancialDetailsServiceISpec.scala [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+          sapDocumentNumberItem = Some("XM00"),
+          chargeReference = Some("XM002610011594"),
+          mainTransaction = Some("1234"),
+          subTransaction = Some("5678"),
+          clearedAmount = Some(123.45),
+          accruedInterest = Some(543.21)
+        )
+      ))
+    )
+    s"call the connector and return a successful result" in {
```
Comment (LewisMitchell) [inline it/services/GetFinancialDetailsServiceISpec.scala:114]:
```suggestion
    "call the connector and return a successful result" in {
```

File: test/models/v3/getPenaltyDetails/FailureResponseSpec.scala [hunk 4]
```diff
@@ -28,7 +30,7 @@ class FailureResponseSpec extends SpecBase {
       |}
       |""".stripMargin)
 
-  val model: FailureResponse = FailureResponse(
+  val model: FailureResponse = v3.FailureResponse(
```
Comment (LewisMitchell) [inline test/models/v3/getPenaltyDetails/FailureResponseSpec.scala:33]: This file needs to be moved to the v3 package

**Review by LewisMitchell — CHANGES_REQUESTED** [review]
Could you also add a test for 404 (non empty body) in the GetFinancialDetailsParserSpec

File: app/connectors/v3/getFinancialDetails/GetFinancialDetailsConnector.scala (hunk 1 above)
Comment (LewisMitchell) [reply app/connectors/v3/getFinancialDetails/GetFinancialDetailsConnector.scala:31]: Probably worth adding it in AppConfig. I believe the params should be:
includeLocks = false
includeStatistical = true
onlyOpenItems = false
calculateAccruedInterest = true
removePOA = false
customerPaymentInformation = false
dateFrom = LocalDate.now().minusYears(2)
dateTo = LocalDate.now()

For includeLocks, includeStatistical, onlyOpenItems, calculateAccruedInterest, removePOA and customerPaymentInformation it may be worth adding them in application.conf just in case our assumption is wrong or more data is needed later.

For dateFrom and dateTo, you may need to update DateHelper to have a method for LocalDate.now

Also worth adding headers as well see: https://github.com/hmrc/penalties/blob/855be3e9dccb6de482d6a22ad4523bba7635dead/app/connectors/v3/getPenaltyDetails/GetPenaltyDetailsConnector.scala#L33

File: test/models/v3/getPenaltyDetails/FailureResponseSpec.scala (hunk 4 above)
Comment (JBlacklaws) [reply test/models/v3/getPenaltyDetails/FailureResponseSpec.scala:33]: FailureResponse is already in the v3 package?

File: test/models/v3/getPenaltyDetails/FailureCodeEnumSpec.scala [hunk 5]
```diff
@@ -17,6 +17,7 @@
 package models.v3.getPenaltyDetails
 
 import base.SpecBase
+import models.v3.FailureCodeEnum
 import play.api.libs.json.{JsString, Json}
```
Comment (LewisMitchell) [inline test/models/v3/getPenaltyDetails/FailureCodeEnumSpec.scala:21]: This also needs to be moved to v3, as it's been moved from /v3/getPenaltyDetails

File: it/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorISpec.scala [hunk 6]
```diff
@@ -26,15 +28,16 @@ import utils.{ETMPWiremock, IntegrationSpecCommonBase}
...
   }
 
   "getFinancialDetails" should {
     "return a successful response when called" in new Setup {
       enableFeatureSwitch(CallAPI1811ETMP)
-      mockResponseForGetFinancialDetailsv3(Status.OK, "VRN/123456789/VATC")
-      val result: GetFinancialDetailsResponse = await(connector.getFinancialDetails("/VRN/123456789/VATC")(hc))
+      mockResponseForGetFinancialDetailsv3(Status.OK,
+        s"VRN/123456789/VATC?dateFrom=${LocalDate.now()}&dateTo=${LocalDate.now()}&onlyOpenItems=false&includeStatistical=true&includeLocks=false&calculateAccruedInterest=true&removePOA=false&customerPaymentInformation=false")
+      val result: GetFinancialDetailsResponse = await(connector.getFinancialDetails("/VRN/123456789/VATC", LocalDate.now(), LocalDate.now())(hc))
```
Comment (LewisMitchell) [inline it/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorISpec.scala:40]: Aren't we just passing in the VRN to the connector?

File: it/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorISpec.scala [hunk 7]
```diff
@@ -47,64 +50,64 @@ class GetFinancialDetailsConnectorISpec extends IntegrationSpecCommonBase with E
...
       result.isLeft shouldBe true
       result.left.get.asInstanceOf[GetFinancialDetailsFailureResponse].status shouldBe Status.INTERNAL_SERVER_ERROR
     }
 
     s"return a $GetFinancialDetailsFailureResponse when the response status is ISE (${Status.SERVICE_UNAVAILABLE})" in new Setup {
       enableFeatureSwitch(CallAPI1811ETMP)
-      mockResponseForGetFinancialDetailsv3(Status.SERVICE_UNAVAILABLE, "VRN/123456789/VATC")
-      val result: GetFinancialDetailsResponse = await(connector.getFinancialDetails("/VRN/123456789/VATC"))
+      mockResponseForGetFinancialDetailsv3(Status.SERVICE_UNAVAILABLE, s"VRN/123456789/VATC?dateFrom=${LocalDate.now()}&dateTo=${LocalDate.now()}&onlyOpenItems=false&includeStatistical=true&includeLocks=false&calculateAccruedInterest=true&removePOA=false&customerPaymentInformation=false")
+      val result: GetFinancialDetailsResponse = await(connector.getFinancialDetails("/VRN/123456789/VATC", LocalDate.now(), LocalDate.now()))
```
Comment (LewisMitchell) [inline it/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorISpec.scala:70]: Aren't we just passing in the VRN to the connector?

File: it/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorISpec.scala [hunk 8]
```diff
@@ -47,64 +50,64 @@ class GetFinancialDetailsConnectorISpec extends IntegrationSpecCommonBase with E
...
       result.isLeft shouldBe true
       result.left.get.asInstanceOf[GetFinancialDetailsFailureResponse].status shouldBe Status.SERVICE_UNAVAILABLE
     }
 
     s"return a $GetFinancialDetailsFailureResponse when the response status is NOT FOUND (${Status.NOT_FOUND})" in new Setup {
       enableFeatureSwitch(CallAPI1811ETMP)
-      mockResponseForGetFinancialDetailsv3(Status.NOT_FOUND, "VRN/123456789/VATC")
-      val result: GetFinancialDetailsResponse = await(connector.getFinancialDetails("/VRN/123456789/VATC"))
+      mockResponseForGetFinancialDetailsv3(Status.NOT_FOUND, s"VRN/123456789/VATC?dateFrom=${LocalDate.now()}&dateTo=${LocalDate.now()}&onlyOpenItems=false&includeStatistical=true&includeLocks=false&calculateAccruedInterest=true&removePOA=false&customerPaymentInformation=false")
+      val result: GetFinancialDetailsResponse = await(connector.getFinancialDetails("/VRN/123456789/VATC", LocalDate.now(), LocalDate.now()))
```
Comment (LewisMitchell) [inline it/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorISpec.scala:78]: Aren't we just passing in the VRN to the connector?

File: it/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorISpec.scala [hunk 9]
```diff
@@ -47,64 +50,64 @@ class GetFinancialDetailsConnectorISpec extends IntegrationSpecCommonBase with E
...
       result.isLeft shouldBe true
       result.left.get.asInstanceOf[GetFinancialDetailsFailureResponse].status shouldBe Status.NOT_FOUND
     }
 
     s"return a $GetFinancialDetailsFailureResponse when the response status is CONFLICT (${Status.CONFLICT})" in new Setup {
       enableFeatureSwitch(CallAPI1811ETMP)
-      mockResponseForGetFinancialDetailsv3(Status.CONFLICT, "VRN/123456789/VATC")
-      val result: GetFinancialDetailsResponse = await(connector.getFinancialDetails("/VRN/123456789/VATC"))
+      mockResponseForGetFinancialDetailsv3(Status.CONFLICT, s"VRN/123456789/VATC?dateFrom=${LocalDate.now()}&dateTo=${LocalDate.now()}&onlyOpenItems=false&includeStatistical=true&includeLocks=false&calculateAccruedInterest=true&removePOA=false&customerPaymentInformation=false")
+      val result: GetFinancialDetailsResponse = await(connector.getFinancialDetails("/VRN/123456789/VATC", LocalDate.now(), LocalDate.now()))
```
Comment (LewisMitchell) [inline it/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorISpec.scala:86]: Aren't we just passing in the VRN to the connector?

File: it/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorISpec.scala [hunk 10]
```diff
@@ -47,64 +50,64 @@ class GetFinancialDetailsConnectorISpec extends IntegrationSpecCommonBase with E
...
       result.isLeft shouldBe true
       result.left.get.asInstanceOf[GetFinancialDetailsFailureResponse].status shouldBe Status.CONFLICT
     }
 
     s"return a $GetFinancialDetailsFailureResponse when the response status is UNPROCESSABLE ENTITY (${Status.UNPROCESSABLE_ENTITY})" in new Setup {
       enableFeatureSwitch(CallAPI1811ETMP)
-      mockResponseForGetFinancialDetailsv3(Status.UNPROCESSABLE_ENTITY, "VRN/123456789/VATC")
-      val result: GetFinancialDetailsResponse = await(connector.getFinancialDetails("/VRN/123456789/VATC"))
+      mockResponseForGetFinancialDetailsv3(Status.UNPROCESSABLE_ENTITY, s"VRN/123456789/VATC?dateFrom=${LocalDate.now()}&dateTo=${LocalDate.now()}&onlyOpenItems=false&includeStatistical=true&includeLocks=false&calculateAccruedInterest=true&removePOA=false&customerPaymentInformation=false")
+      val result: GetFinancialDetailsResponse = await(connector.getFinancialDetails("/VRN/123456789/VATC", LocalDate.now(), LocalDate.now()))
```
Comment (LewisMitchell) [inline it/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorISpec.scala:94]: Aren't we just passing in the VRN to the connector?

File: it/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorISpec.scala [hunk 11]
```diff
@@ -47,64 +50,64 @@ class GetFinancialDetailsConnectorISpec extends IntegrationSpecCommonBase with E
...
       result.isLeft shouldBe true
       result.left.get.asInstanceOf[GetFinancialDetailsFailureResponse].status shouldBe Status.UNPROCESSABLE_ENTITY
     }
 
     s"return a $GetFinancialDetailsFailureResponse when the response status is BAD REQUEST (${Status.BAD_REQUEST})" in new Setup {
       enableFeatureSwitch(CallAPI1811ETMP)
-      mockResponseForGetFinancialDetailsv3(Status.BAD_REQUEST, "VRN/123456789/VATC")
-      val result: GetFinancialDetailsResponse = await(connector.getFinancialDetails("/VRN/123456789/VATC"))
+      mockResponseForGetFinancialDetailsv3(Status.BAD_REQUEST, s"VRN/123456789/VATC?dateFrom=${LocalDate.now()}&dateTo=${LocalDate.now()}&onlyOpenItems=false&includeStatistical=true&includeLocks=false&calculateAccruedInterest=true&removePOA=false&customerPaymentInformation=false")
+      val result: GetFinancialDetailsResponse = await(connector.getFinancialDetails("/VRN/123456789/VATC", LocalDate.now(), LocalDate.now()))
```
Comment (LewisMitchell) [inline it/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorISpec.scala:102]: Aren't we just passing in the VRN to the connector?

File: it/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorISpec.scala [hunk 12]
```diff
@@ -47,64 +50,64 @@ class GetFinancialDetailsConnectorISpec extends IntegrationSpecCommonBase with E
...
       result.isLeft shouldBe true
       result.left.get.asInstanceOf[GetFinancialDetailsFailureResponse].status shouldBe Status.BAD_REQUEST
     }
 
     s"return a $GetFinancialDetailsFailureResponse when the response status is FORBIDDEN (${Status.FORBIDDEN})" in new Setup {
       enableFeatureSwitch(CallAPI1811ETMP)
-      mockResponseForGetFinancialDetailsv3(Status.FORBIDDEN, "VRN/123456789/VATC")
-      val result: GetFinancialDetailsResponse = await(connector.getFinancialDetails("/VRN/123456789/VATC"))
+      mockResponseForGetFinancialDetailsv3(Status.FORBIDDEN, s"VRN/123456789/VATC?dateFrom=${LocalDate.now()}&dateTo=${LocalDate.now()}&onlyOpenItems=false&includeStatistical=true&includeLocks=false&calculateAccruedInterest=true&removePOA=false&customerPaymentInformation=false")
+      val result: GetFinancialDetailsResponse = await(connector.getFinancialDetails("/VRN/123456789/VATC", LocalDate.now(), LocalDate.now()))
```
Comment (LewisMitchell) [inline it/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorISpec.scala:110]: Aren't we just passing in the VRN to the connector?

File: test/connectors/parsers/v2/GetFinancialDetailsParserSpec.scala [hunk 13]
```diff
@@ -44,6 +44,16 @@ class GetFinancialDetailsParserSpec extends AnyWordSpec with Matchers {
...
     ), headers = Map.empty)
+  val mockNotFoundHttpResponseWithBody: HttpResponse = HttpResponse.apply(status = Status.NOT_FOUND, json = Json.parse(
+    """
+      |{
+      | "documentDetails": [{
+      |   "summary": {
+      |     }
+      |   }]
+      | }
+      |""".stripMargin
```
Comment (LewisMitchell) [inline test/connectors/parsers/v2/GetFinancialDetailsParserSpec.scala:55]: If it were a 404, I don't think the body would look like this, instead it would look like this:
```
        """
          |{
          | "failures":[
          |   {
          |     "code": "NO_DATA_FOUND",
          |     "reason": "This is a reason"
          |   }
          | ]
          |}
```

File: test/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorSpec.scala [hunk 14]
```diff
@@ -27,20 +27,26 @@ import org.mockito.Mockito._
...
+class GetFinancialDetailsConnectorSpec extends SpecBase {
   implicit val ec: ExecutionContext = ExecutionContext.Implicits.global
   val mockHttpClient: HttpClient = mock(classOf[HttpClient])
   val mockAppConfig: AppConfig = mock(classOf[AppConfig])
+  val mockDateHelper: DateHelper = mock(classOf[DateHelper])
+  val sampleDate: LocalDate = LocalDate.of(2023, 1, 1)
 
   class Setup {
     reset(mockHttpClient)
     reset(mockAppConfig)
```
Comment (LewisMitchell) [inline test/connectors/v3/getFinancialDetails/GetFinancialDetailsConnectorSpec.scala:43]: Could you add a reset(mockDateHelper) here please
```suggestion
    reset(mockAppConfig)
    reset(mockDateHelper)
```

File: test/services/GetFinancialDetailsServiceSpec.scala [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+    s"call the connector and return a $GetFinancialDetailsSuccessResponse when the request is successful" in new Setup {
+      when(mockGetFinancialDetailsConnector.getFinancialDetails(Matchers.eq("123456789"), Matchers.eq(LocalDate.now()), Matchers.eq(LocalDate.now()))(any()))
+        .thenReturn(Future.successful(Right(GetFinancialDetailsSuccessResponse(mockGetFinancialDetailsResponseAsModel))))
+
+      val result: GetFinancialDetailsResponse = await(service.getDataFromFinancialServiceForVATVCN("123456789", LocalDate.now(), LocalDate.now()))
+      result.isRight shouldBe true
+      result.right.get shouldBe GetFinancialDetailsSuccessResponse(mockGetFinancialDetailsResponseAsModel)
+    }
+  }
```
Comment (LewisMitchell) [inline test/services/GetFinancialDetailsServiceSpec.scala:137]: Is it worth adding some more tests here? Failure scenarios? I've seen in the IT's that this is the case but it may be worth adding unit tests for them as well?

**Review by LewisMitchell — CHANGES_REQUESTED** [review]
Looks good, mainly just test tweaks/additions
